### ISSUE TITLE
4.0.0

### ISSUE THREAD

**Issue by jjatie** [danielgindi/Charts#3134]
I think we should set some goals for 4.0.0.

I am adding things miscellaneously, but we should set some concrete targets. A few things I would like to see:

1. The most major thing I would like to see is protocol oriented ChartData/DataSet/Entry instead of subclassing. This would allow us to more easily move to value types when we eventually move to a pure Swift library. I experimented with this before, thinking it would result in significantly more code, but it turned out to be significantly less (I don't have line counts, but it's 153kb vs 123kb of the current ChartData/DataSet/Entry vs a protocol oriented one using structs).

2. I would like to figure out a way to have collection conformance added to ChartDataSets as it provides a more "Swift" interface and would allow us to greatly reduce our code base.

3. I would like to reevaluate what should and shouldn't be made public. There are more than a few areas where it seems like things have unnecessarily open access control. The more we keep internal, the more we are able to make things pure Swift. I understand we are trying to stay in line with MPAndroidCharts, but in my experience Java likes to just keep EVERYTHING open.

4. I would like to reevaluate naming. There are many spots where we deviate from the Swift API guidelines. I'm guessing it's just because we started at Swift 1 and things have changed since then. It would be great to see it uniform.

5. Finally, I would like us to explore using Core Animation. If it is indeed more performant, this would be a great feature to advertise in 4.0 and would bring us one step closer to making for-pay chart frameworks unnecessary.

**Comment by liuxuan30** [comment]
for 3, the context is we used to have more restricted access control, but people are asking they want to subclass and override outside the module, so we made almost members to be open/public.
for 5, I think it's beyond 4.0. 

In my opinion, my 4.0 goal is to finish all your swift refactoring, alone with some bugfix, and it takes time for our users to adopt and change their code base.

Nice to have is we could carry out gradient line and naming axis (the other two projects)
